File: source/javascript/test_runner.js
// Mini test runner framework
const tests = [];
let passed = 0, failed = 0;

function addTest(name, actual, expected) { tests.push({ name, actual, expected }); }

function runTests() {
    console.log("=== Test Results ===");
    for (const t of tests) {
        if (t.actual === t.expected) { console.log(`  PASS: ${t.name}`); passed++; }
        else { console.log(`  FAIL: ${t.name} (expected ${t.expected}, got ${t.actual})`); failed++; }
    }
    const total = passed + failed;
    console.log(`\n${passed}/${total} tests passed`);
    if (failed > 0) console.log(`${failed} test(s) failed`);
}

// String tests
addTest("string length", "hello".length, 5);
addTest("string contains", "hello world".includes("world"), true);
addTest("string upper", "hello".toUpperCase(), "HELLO");

// Array tests
const arr = [1, 2, 3, 4, 5];
addTest("array length", arr.length, 5);
addTest("array first", arr[0], 1);
addTest("array last", arr[arr.length - 1], 5);
addTest("array contains", arr.includes(3), true);

// Math tests
addTest("addition", 2 + 3, 5);
addTest("multiplication", 6 * 7, 42);
addTest("boolean logic", true && !false, true);

runTests();

File: source/rust/test_runner.rs
struct Test {
    name: String,
    passed: bool,
    expected: String,
    actual: String,
}

fn run_tests(tests: &[Test]) {
    println!("=== Test Results ===");
    let mut passed = 0;
    let mut failed = 0;
    for t in tests {
        if t.passed {
            println!("  PASS: {}", t.name);
            passed += 1;
        } else {
            println!("  FAIL: {} (expected {}, got {})", t.name, t.expected, t.actual);
            failed += 1;
        }
    }
    let total = passed + failed;
    println!("\n{}/{} tests passed", passed, total);
    if failed > 0 {
        println!("{} test(s) failed", failed);
    }
}

fn test_eq<T: PartialEq + ToString>(name: &str, actual: T, expected: T) -> Test {
    let passed = actual == expected;
    Test {
        name: name.to_string(),
        passed,
        expected: expected.to_string(),
        actual: actual.to_string(),
    }
}

fn main() {
    let hello = "hello";
    let arr = vec![1i32, 2, 3, 4, 5];

    let tests = vec![
        test_eq("string length", hello.len(), 5),
        test_eq("string contains", hello.contains("world"), false), // "hello world".contains("world")
        {
            // Fix: use the right string
            let s = "hello world";
            test_eq("string contains", s.contains("world"), true)
        },
        test_eq("string upper", hello.to_uppercase(), "HELLO".to_string()),
        test_eq("array length", arr.len(), 5),
        test_eq("array first", arr.first().copied().unwrap(), 1),
        test_eq("array last", arr.last().copied().unwrap(), 5),
        test_eq("array contains", arr.contains(&3), true),
        test_eq("addition", 2 + 3, 5),
        test_eq("multiplication", 6 * 7, 42),
        test_eq("boolean logic", true && !false, true),
    ];

    // Filter out duplicates - the Hemlock code has 11 tests
    // Re-create to match: string length, string contains, string upper, array x4, math x2, bool
    let tests2 = vec![
        test_eq("string length", "hello".len(), 5usize),
        test_eq("string contains", "hello world".contains("world"), true),
        test_eq("string upper", "hello".to_uppercase(), "HELLO".to_string()),
        test_eq("array length", arr.len(), 5usize),
        test_eq("array first", *arr.first().unwrap(), 1i32),
        test_eq("array last", *arr.last().unwrap(), 5i32),
        test_eq("array contains", arr.contains(&3), true),
        test_eq("addition", 2 + 3, 5i32),
        test_eq("multiplication", 6 * 7, 42i32),
        test_eq("boolean logic", true && !false, true),
    ];

    run_tests(&tests2);
}
